File: 1301-1400/1345-Jump Game IV/1345-Jump Game IV.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        unordered_map<int, vector<int>> table;
        for (int i = 0; i < arr.size(); ++i) {
            table[arr[i]].push_back(i);
        }
        
        vector<bool> visited(arr.size());
        queue<int> Q;
        Q.push(0);
        visited[0] = true;
        int step = 0;
        while (!Q.empty()) {
            int qs = Q.size();
            for (int i = 0; i < qs; ++i) {
                int curr = Q.front();
                if (curr == arr.size() - 1) {
                    return step;
                }
                Q.pop();
                if (curr - 1 >= 0 && !visited[curr - 1]) {
                    Q.push(curr - 1);
                    visited[curr - 1] = true;
                }
                if (curr + 1 < arr.size() && !visited[curr + 1]) {
                    Q.push(curr + 1);
                    visited[curr + 1] = true;
                }
                for (int next : table[arr[curr]]) {
                    if (!visited[next]) {
                        Q.push(next);
                        visited[next] = true;
                    }
                }
                table[arr[curr]].clear();
            }
            ++step;
        }
        
        return step;
    }
};

File: 1301-1400/1345-Jump Game IV/1345-Jump Game IV.py
import collections
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        table = collections.defaultdict(list)
        for i, a in enumerate(arr):
            table[a].append(i)
        visited = [False] * len(arr)
        Q = collections.deque([0])
        visited[0] = True
        step = 0
        while Q:
            for _ in range(len(Q)):
                curr = Q.popleft()
                if curr == len(arr) - 1:
                    return step
                if curr - 1 >= 0 and not visited[curr - 1]:
                    Q.append(curr - 1)
                    visited[curr - 1] = True
                if curr + 1 < len(arr) and not visited[curr + 1]:
                    Q.append(curr + 1)
                    visited[curr + 1] = True
                for next in table[arr[curr]]:
                    if not visited[next]:
                        Q.append(next)
                        visited[next] = True
                table[arr[curr]].clear()
            step += 1
        return step
